### PR TITLE
[BEAM-12970][Playground] Implement gRPC API - RunCode

### PR DESCRIPTION
**Pull request by AydarZaynutdinov**
Implement the `RunCode` method.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: playground/backend/cmd/server/controller.go [hunk 1]
```diff
@@ -16,19 +16,77 @@ package main
...
 }
 
+var cacheService = cache.NewCache(os.Getenv("cache"))
+
 //RunCode is running code from requests using a particular SDK
 func (controller *playgroundController) RunCode(ctx context.Context, info *pb.RunCodeRequest) (*pb.RunCodeResponse, error) {
-	// TODO implement this method
-	pipelineInfo := pb.RunCodeResponse{PipelineUuid: uuid.NewString()}
+	defer func() {
+		log.Println("RunCode complete")
```
Comment (ilya-kozyrev) [inline playground/backend/cmd/server/controller.go:40]: Do we really want to log each interaction with our service just as "RunCode complete"?

File: playground/backend/cmd/server/controller.go [hunk 2]
```diff
@@ -16,19 +16,77 @@ package main
...
+	}()
+
+	pipelineId := uuid.New()
+	env := environment.NewEnvironment()
```
Comment (ilya-kozyrev) [inline playground/backend/cmd/server/controller.go:44]: Why did you create a new environment here? That's static at the moment when the server is starting. Just keep it as an argument in the controller constructor.

File: playground/backend/cmd/server/controller.go [hunk 3]
```diff
@@ -53,3 +111,183 @@ func (controller *playgroundController) GetCompileOutput(ctx context.Context, in
 	compileOutput := pb.GetCompileOutputResponse{Output: "test compile output"}
 	return &compileOutput, nil
 }
+
+// runCode validates, compiles and runs code by pipelineId.
+// During each operation updates status of execution and saves it into cache.
+// In case of compilation failed saves logs to cache.
+// After success code running saves output to cache.
+func runCode(ctx context.Context, lc *fs_tool.LifeCycle, exec *executors.Executor, pipelineId uuid.UUID, timeout time.Duration) {
```
Comment (ilya-kozyrev) [inline playground/backend/cmd/server/controller.go:119]: You keep doing that :) Similarly named methods with different implementation. The first capital letter says nothing about its functionality.

File: playground/backend/cmd/server/controller.go [hunk 4]
```diff
@@ -53,3 +111,183 @@ func (controller *playgroundController) GetCompileOutput(ctx context.Context, in
...
+	log.Println("parallelRunCode: DeleteFolders ...")
+	err := lc.DeleteFolders()
+	if err != nil {
+		log.Printf("parallelRunCode: DeleteFolders: " + err.Error())
+	}
+	log.Println("parallelRunCode: DeleteFolders complete")
+}
+
+// processErrDuringValidate processes error received during Validate step
+func processErrDuringValidate(err error, pipelineId uuid.UUID) {
```
Comment (ilya-kozyrev) [inline playground/backend/cmd/server/controller.go:223]: How do you plan test functions like this?

File: playground/backend/internal/environment/environment_service.go [hunk 5]
```diff
@@ -54,6 +59,9 @@ func NewEnvironment() *Environment {
 func getServerEnvsFromOsEnvs() *ServerEnvs {
 	ip := defaultIp
 	port := defaultPort
+	cacheExpiration := defaultCacheExpiration
+	runCodeTimeout := defaultRunCodeTimeout
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/environment_service.go:63]: code/pipeline execution timeout

File: playground/backend/internal/environment/network.go [hunk 6]
```diff
@@ -15,20 +15,35 @@
...
+	"time"
+)
 
 // ServerEnvs contains all environment variables that need to run server.
 type ServerEnvs struct {
-	ip   string
-	port int
+	ip              string
+	port            int
+	cacheExpiration time.Duration
```
Comment (ilya-kozyrev) [inline playground/backend/internal/environment/network.go:27]: How it's related to the network?

**Review by ilya-kozyrev — CHANGES_REQUESTED** [review]

File: playground/backend/cmd/server/controller.go [hunk 7]
```diff
@@ -53,3 +111,183 @@ func (controller *playgroundController) GetCompileOutput(ctx context.Context, in
...
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+	}
+
+	// set to cache pipelineId: cache.SubKey_Status: pb.Status_STATUS_ERROR
+	err = cacheService.SetValue(pipelineId, cache.SubKey_Status, pb.Status_STATUS_ERROR)
+	if err != nil {
+		log.Printf("parallelRunCode: cache.SetValue: " + err.Error())
+	}
+}
```
Comment (KhaninArtur) [inline playground/backend/cmd/server/controller.go:276]: I see too many common actions in these methods – can we create a generic function for such processing?

File: playground/backend/cmd/server/controller.go [hunk 8]
```diff
@@ -53,3 +111,183 @@ func (controller *playgroundController) GetCompileOutput(ctx context.Context, in
...
+	err := lc.DeleteFolders()
+	if err != nil {
+		log.Printf("parallelRunCode: DeleteFolders: " + err.Error())
+	}
+	log.Println("parallelRunCode: DeleteFolders complete")
+}
+
+// processErrDuringValidate processes error received during Validate step
+func processErrDuringValidate(err error, pipelineId uuid.UUID) {
+	log.Printf("parallelRunCode: Validate: " + err.Error())
```
Comment (KhaninArtur) [inline playground/backend/cmd/server/controller.go:224]: Let's add `pipelineId` for these logs too.

File: playground/backend/cmd/server/controller_test.go [hunk 9]
```diff
@@ -54,23 +55,67 @@ func teardown(server *grpc.Server) {
...
+			}
+			if err == nil {
+				if response == nil {
+					t.Errorf("PlaygroundController_RunCode() response shoudn't be nil")
+				} else {
+					if response.PipelineUuid == "" {
+						t.Errorf("PlaygroundController_RunCode() response.pipeLineId shoudn't be nil")
+					} else {
+						path := "internal/executable_files_" + response.PipelineUuid
+						//absPath, _ := filepath.Abs(path)
```
Comment (KhaninArtur) [inline playground/backend/cmd/server/controller_test.go:112]: Could you please remove this "saved" code?😁

File: playground/backend/internal/environment/network.go [hunk 10]
```diff
@@ -15,20 +15,35 @@
...
+	runCodeTimeout  time.Duration
 }
 
 // NewServerEnvs constructor for ServerEnvs
-func NewServerEnvs(ip string, port int) *ServerEnvs {
-	return &ServerEnvs{ip: ip, port: port}
+func NewServerEnvs(ip string, port int, cacheExpiration, runCodeTimeout time.Duration) *ServerEnvs {
```
Comment (KhaninArtur) [inline playground/backend/internal/environment/network.go:32]: Shouldn't we provide a type for `cacheExpiration`?

File: playground/backend/cmd/server/controller.go (hunk 1 above)
Comment (ilya-kozyrev) [reply playground/backend/cmd/server/controller.go:40]: Yep, agree. One minor comment - don't use method name; let's use smth like "pipeline {id} started"

File: playground/backend/cmd/server/controller.go (hunk 2 above)
Comment (AydarZaynutdinov) [reply playground/backend/cmd/server/controller.go]: Changed.

File: playground/backend/cmd/server/controller.go (hunk 3 above)
Comment (AydarZaynutdinov) [reply playground/backend/cmd/server/controller.go]: Changed.

File: playground/backend/internal/environment/network.go (hunk 6 above)
Comment (AydarZaynutdinov) [reply playground/backend/internal/environment/network.go]: Changed.

File: playground/backend/cmd/server/controller.go (hunk 8 above)
Comment (AydarZaynutdinov) [reply playground/backend/cmd/server/controller.go]: Changed.

File: playground/backend/cmd/server/controller_test.go (hunk 9 above)
Comment (AydarZaynutdinov) [reply playground/backend/cmd/server/controller_test.go]: Chaned.

**Pull request closed without merging**